crates/sui-data-ingestion/src/workers/archival.rs

```diff
@@ -0,0 +1,189 @@
+            SUMMARY_FILE_MAGIC,
+            &state.summary_buffer,
+        )
+        .await?;
+        let manifest_bytes = self
+            .remote_store
+            .get(&Path::from("MANIFEST"))
+            .await?
+            .bytes()
+            .await?;
+        let mut manifest = read_manifest_from_bytes(manifest_bytes.to_vec())?;
+        let checkpoint_file_metadata = create_file_metadata(
+            std::path::Path::new(&checkpoint_file_path),
+            FileType::CheckpointContent,
+            state.epoch,
+            state.checkpoint_range.clone(),
+        )?;
+        let summary_file_metadata = create_file_metadata(
+            std::path::Path::new(&summary_file_path),
+            FileType::CheckpointSummary,
+            state.epoch,
+            state.checkpoint_range.clone(),
+        )?;
+        manifest.update(
+            state.epoch,
+            state.checkpoint_range.start,
+            checkpoint_file_metadata,
+            summary_file_metadata,
+        );
+        let bytes = finalize_manifest(manifest)?;
+        self.remote_store
+            .put(&Path::from("MANIFEST"), bytes)
+            .await?;
+        Ok(())
+    }
+
+    async fn upload_file(&self, location: Path, magic: u32, content: &[u8]) -> Result<()> {
+        let mut buffer = vec![0; 4];
+        BigEndian::write_u32(&mut buffer, magic);
+        buffer.push(StorageFormat::Blob.into());
+        buffer.push(FileCompression::Zstd.into());
+        buffer.extend_from_slice(content);
+        let mut compressed_buffer = vec![];
+        let mut cursor = Cursor::new(buffer);
+        compress(&mut cursor, &mut compressed_buffer)?;
+        self.remote_store
+            .put(&location, Bytes::from(compressed_buffer))
+            .await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Worker for ArchivalWorker {
+    async fn process_checkpoint(&self, checkpoint: CheckpointData) -> Result<()> {
+        let mut state = self.state.lock().await;
+        let sequence_number = checkpoint.checkpoint_summary.sequence_number;
+        let epoch = checkpoint.checkpoint_summary.epoch;
+        if state.checkpoint_range.is_empty() {
+            state.reset(epoch, sequence_number);
+        }
+        let full_checkpoint_contents = FullCheckpointContents::from_contents_and_execution_data(
+            checkpoint.checkpoint_contents,
+            checkpoint
+                .transactions
+                .iter()
+                .map(|t| ExecutionData::new(t.transaction.clone(), t.effects.clone())),
+        );
+        let contents_blob = Blob::encode(&full_checkpoint_contents, BlobEncoding::Bcs)?;
+        if ((state.buffer.len() + contents_blob.size()) > self.commit_file_size)
+            || state.last_commit_instant.elapsed() > self.commit_duration
+            || state.epoch != epoch
+        {
+            self.upload(&state).await?;
+            state.reset(epoch, sequence_number);
+            state.should_update_progress = true;
```

Comment:
How are we maintaining consistency in case of process restart after updating manifest and before saving the progress state. Is this going to start from the next checkpoint after the last one in manifest?

Reply:
good question, if process restarts after it basically saved everything but haven't updated the meta store, it will start again from the last known watermark in the meta store. Is that an issue? (`chk` files will get rewritten)  tbh not fully aware of required guarantees for the manifest file

Reply:
added manifest read in the constructor and a check in `process_checkpoint`. Now in the following scenario, all checkpoints will be technically replayed, but ignored by the worker